src/deployer.ts

```diff
@@ -0,0 +1,194 @@
+    .choices('local', intersection)
+    .demandOption('local')
+    .array('remotes')
+    .describe(
+      'remotes',
+      "The chains with which 'local' will be able to send and receive messages",
+    )
+    .choices('remotes', intersection)
+    .demandOption('remotes')
+    .describe('key', 'A hexadecimal private key for transaction signing')
+    .string('key')
+    .coerce('key', assertBytes32)
+    .demandOption('key')
+    .middleware(
+      assertBalances(multiProvider, (argv) => argv.remotes.concat(argv.local)),
+    )
+    .describe('write-agent-config', 'Whether or not to write agent config')
+    .default('write-agent-config', true)
+    .boolean('write-agent-config').argv;
+}
+
+type MultisigIsmContracts = {
+  multisigIsm: LegacyMultisigIsm;
+};
+
+export class HyperlanePermissionlessDeployer {
+  constructor(
+    protected multiProvider: MultiProvider,
+    protected signer: ethers.Signer,
+    protected local: ChainName,
+    protected remotes: ChainName[],
+    protected writeAgentConfig?: boolean,
+  ) {}
+
+  static async fromArgs(): Promise<HyperlanePermissionlessDeployer> {
+    const multiProvider = getMultiProvider();
+    const { local, remotes, key, writeAgentConfig } = await getArgs(
+      multiProvider,
+    );
+    const signer = new ethers.Wallet(key);
+    multiProvider.setSharedSigner(signer);
+
+    return new HyperlanePermissionlessDeployer(
+      multiProvider,
+      signer,
+      local,
+      remotes as unknown as string[],
+      writeAgentConfig,
+    );
+  }
+
+  get chains(): ChainName[] {
+    return this.remotes.concat([this.local]);
+  }
+
+  async deploy(): Promise<void> {
+    let contracts: ChainMap<HyperlaneContracts> = {};
+    const owner = await this.signer.getAddress();
+    // First, deploy core contracts to the local chain
+    // NB: We create core configs for *all* chains because
+    // we also use coreDeployer to deploy MultisigIsms.
+    // Once we move that out to a HyperlaneIsmDeployer
+    // we can just do:
+    // const coreContracts = await coreDeployer.deploy();
+    const coreConfig = buildCoreConfig(owner, this.chains);
+    const coreDeployer = new HyperlaneCoreDeployer(
+      this.multiProvider,
+      coreConfig,
+    );
+    const coreContracts: ChainMap<CoreContracts> = {};
+    coreContracts[this.local] = await coreDeployer.deployContracts(
+      this.local,
+      coreConfig[this.local],
+    );
+    contracts = objMerge(contracts, coreContracts);
+
+    // Next, deploy MultisigIsms to the remote chains
+    // TODO: Would be cleaner if using HyperlaneIsmDeployer
+    const isms: ChainMap<MultisigIsmContracts> = {};
+    isms[this.local] = {
+      multisigIsm: coreContracts[this.local].multisigIsm,
+    };
+    for (const remote of this.remotes) {
+      isms[remote] = {
+        multisigIsm: await coreDeployer.deployLegacyMultisigIsm(remote),
+      };
+    }
+    contracts = objMerge(contracts, isms);
+
+    // Next, deploy TestRecipients to all chains
```

Comment:
should this be optional?

Reply:
Maybe in the future, but for now I think having people validate their agents are running properly is an important step in the deployment process